apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-can-preview-but-need-upgrade.ts

```diff
@@ -55,13 +63,137 @@ const checkNeedUpgrade = ( {
 	return true;
 };
 
+/**
+ * A polling mechanism to check whether the user has upgraded the plan.
+ */
+const usePolling = ( {
+	isPolling,
+	previewingTheme,
+	setCanPreviewButNeedUpgrade,
+	setIsPolling,
+}: {
+	isPolling: boolean;
+	previewingTheme: ReturnType< typeof usePreviewingTheme >;
+	setCanPreviewButNeedUpgrade: ( canPreviewButNeedUpgrade: boolean ) => void;
+	setIsPolling: ( isPolling: boolean ) => void;
+} ) => {
+	const pollInterval = useRef( POLL_INTERVAL_DEFAULT );
+	const timeoutId = useRef< NodeJS.Timeout | null >( null );
+	const [ visibilityChange, setVisibilityChange ] = useState( false );
+	const [ shouldReflectUpgradedPlan, setShouldReflectUpgradedPlan ] = useState( false );
+	const { createInfoNotice, removeNotice } = useDispatch( 'core/notices' );
+
+	const noticeText = __(
+		'You have successfully upgraded your plan! You can now activate the theme.',
+		'wpcom-live-preview'
+	);
+
+	useEffect( () => {
+		if ( ! isPolling ) {
+			return;
+		}
+		// Using exponential backoff, as we don't want to make too many requests while the user is completing the purchase.
+		const nextPollInterval = () =>
+			Math.min( pollInterval.current * POLL_INTERVAL_FACTOR, POLL_INTERVAL_MAX );
+		const poll = () => {
+			wpcom.req
+				.get( `/sites/${ window._currentSiteId }`, { apiVersion: '1.2' } )
+				.then( ( site: any ) => {
+					if ( ! needUpgrade( { site, previewingThemeType: previewingTheme.type } ) ) {
+						setShouldReflectUpgradedPlan( true );
+						setIsPolling( false );
```

Comment:
It would be better to close the opened window as well. Or the opened window can check whether the `window.opener` exists or not. If so, we can close the window when the site is upgraded successfully. And then we can re-fetch the plan only when the opened window is closed

Reply:
Thank you for your comment!

Users will see kinda "instructions" for the next steps after upgrading their plan. 
<img width="1437" alt="Screenshot 2023-12-27 at 16 40 29" src="https://github.com/Automattic/wp-calypso/assets/5287479/f354fd27-e906-49d6-aa51-40198acb3ad2">
_after clicking `No Thanks` in the above page_
<img width="1429" alt="Screenshot 2023-12-27 at 16 40 44" src="https://github.com/Automattic/wp-calypso/assets/5287479/9aa378a4-2872-4968-a3d2-48353c1b2dff">

So I thought that closing the opened window suddenly might potentially lead to confusion, especially if users need to refer to the instructions. But most instructions might not be necessary for Live Preview flow anyway. 

What do you think? @arthur791004 @Automattic/lego

Reply:
I agree that we should close the opened window, otherwise users might not know that they were supposed to get back to the original window and continue interacting with the opened window.

Reply:
**UPDATE** We will close the opened checkout window when
* The checkout is completed
* The checkout is failed
* The checkout is canceled (Click the back button on the top-left)

Reply:
**Display the info notice when the checkout is completed**

![image](https://github.com/Automattic/wp-calypso/assets/13596067/cfffff02-7394-4246-912f-9320667330ae)

**Display the error notice when the checkout is failed**

![image](https://github.com/Automattic/wp-calypso/assets/13596067/a9d06145-128a-4174-b55d-d2bf3bf70584)